### PR TITLE
feat: multi address

### PR DESCRIPTION
**Pull request by ganchoradkov**
# Breaking Changes

N/A

# Changes
 
- Added ability to see all connected accounts & switch between them  
image
image

# Testing
https://www.notion.so/walletconnect/Bug-Bash-Multi-account-bcd4f8d6d8e94fafac459c77daff194e

### CODE REVIEW

**Comment by lukaisailovic** [comment]
@ganchoradkov Can we please add more context? What was the issue, how was it resolved etc

File: packages/scaffold-ui/src/modal/w3m-router/index.ts [hunk 1]
```diff
@@ -75,6 +75,22 @@ export class W3mRouter extends LitElement {
...
+      case 'SelectAddresses':
+        return html`<w3m-select-addresses-view></w3m-select-addresses-view>`
+      case 'SwitchAddress':
+        return html`<w3m-switch-address-view></w3m-switch-address-view>`
+      case 'WhatIsAWallet':
+        return html`<w3m-what-is-a-wallet-view></w3m-what-is-a-wallet-view>`
+      case 'WhatIsANetwork':
+        return html`<w3m-what-is-a-network-view></w3m-what-is-a-network-view>`
+      case 'GetWallet':
+        return html`<w3m-get-wallet-view></w3m-get-wallet-view>`
```
Comment (enesozturk) [inline packages/scaffold-ui/src/modal/w3m-router/index.ts:93]: Previously it was looking sorted alphabetically, why these are moved different rows?

File: packages/scaffold-ui/src/partials/w3m-account-default-widget/index.ts [hunk 2]
```diff
@@ -77,42 +77,24 @@ export class W3mAccountDefaultWidget extends LitElement {
...
 
     return html`<wui-flex
         flexDirection="column"
         .padding=${['0', 'xl', 'm', 'xl'] as const}
         alignItems="center"
         gap="l"
       >
-        <wui-avatar
-          alt=${ifDefined(this.address)}
+        <wui-profile-button-v2
```
Comment (enesozturk) [inline packages/scaffold-ui/src/partials/w3m-account-default-widget/index.ts:89]: Why don't we refactor the current `wui-profile-button` since it's only used here? We'll be having two different versions unnecessarily. Bc also the `*-v2` naming not quite good/descriptive.

File: packages/scaffold-ui/src/views/w3m-profile-view/index.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    this.usubscribe.push(
+      AccountController.subscribe(val => {
+        if (val.address) {
+          this.address = val.address
+          this.profileImage = val.profileImage
+          this.profileName = val.profileName
+        } else {
+          ModalController.close()
+        }
+      })
```
Comment (enesozturk) [inline packages/scaffold-ui/src/views/w3m-profile-view/index.ts:44]: Can we use `subscribeKey` to not listen unnecessary changes of the AccountController please?
```suggestion
      AccountController.subscribeKey("address", val => {...}),
      AccountController.subscribeKey("profileImage", val => {...}),
      AccountController.subscribeKey("profileName", val => {...}),
```

File: packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    })
+  }
+
+  onSelectAll = (event: Event) => {
+    const checked = (event.target as HTMLInputElement).checked
+    this.selectAll = this.selectedAccounts.length === this.allAccounts.length
+    this.allAccounts.forEach(account => {
+      this.onSelect(account, checked)
+    })
+  }
```
Comment (enesozturk) [inline packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts:40]: Can we make these functions `private` and use named function rather than anonymous functions to follow same convention and also not having hoisting issues please? Also we could move them to bottom (the private methods section)
```suggestion
  private onSelectAll(event: Event) {
    const checked = (event.target as HTMLInputElement).checked
    this.selectAll = this.selectedAccounts.length === this.allAccounts.length
    this.allAccounts.forEach(account => {
      this.onSelect(account, checked)
    })
  }
```

File: packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+  onSelect = (account: AccountType, add: boolean) => {
```
Comment (enesozturk) [inline packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts:42]: Same for `onSelect` this and `getAddressTemplate`

File: packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        .checked="${checked}"
+        @change="${this.handleClick(account)}"
+      />
+    </wui-list-account>`
+  }
+
+  private handleClick = (account: AccountType) => (event: Event) => {
+    const target = event.target as HTMLInputElement
+    this.onSelect?.({ ...account }, target?.checked)
+  }
```
Comment (enesozturk) [inline packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts:114]: Let's avoid anonymous functions.
```suggestion
  private handleClick(account: AccountType) {
  return (event: Event) => {
    const target = event.target as HTMLInputElement;
    this.onSelect?.({ ...account }, target?.checked);
  };
}
```

File: packages/scaffold-ui/src/views/w3m-switch-address-view/index.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      <wui-flex flexDirection="column" gap="xxl" .padding=${['l', 'xl', 'xl', 'xl'] as const}>
+        ${this.allAccounts.map(account => this.getAddressTemplate(account))}
+      </wui-flex>
+    `
+  }
+
+  // -- Private ------------------------------------------- //
+
+  private getAddressTemplate(account: AccountType) {
+    // If there is custom label for the address, use it
```
Comment (enesozturk) [inline packages/scaffold-ui/src/views/w3m-switch-address-view/index.ts:77]: Unless there is a complex logic / calculation, the code shouldn't require comments. The line is quite readable, I think we could remove the comment.

File: packages/ui/src/composites/wui-dropdown-menu/styles.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  :host {
+    z-index: calc(var(--w3m-z-index) + 1);
+    width: 200px;
+    padding: var(--wui-spacing-3xs);
+    align-items: center;
+    display: inherit;
+    border-radius: var(--wui-border-radius-xs);
+    border: 1px solid var(--wui-gray-glass-002);
+    background: var(--wui-color-fg-base-125);
+    /* Dark/Elevation/L */
```
Comment (enesozturk) [inline packages/ui/src/composites/wui-dropdown-menu/styles.ts:13]: Do we need to create CSS variable for this shadow value? Is this `Dark/Elevation/L` coming from the design system? Could you share a link for the designs?

File: packages/ui/src/composites/wui-list-account/index.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  @property() public accountAddress = ''
+
+  @property() public accountType = ''
+
+  private connectedConnector = StorageUtil.getConnectedConnector()
+  private labels = AccountController.state.addressLabels
+  private balance = 0
+  private fetchingBalance = true
+  private shouldShowIcon = false
+  private caipNetwork = NetworkController.state.caipNetwork
```
Comment (enesozturk) [inline packages/ui/src/composites/wui-list-account/index.ts:32]: Can we separate properties and states with new lines as we do in all other comps?
```suggestion
  private connectedConnector = StorageUtil.getConnectedConnector()
  
  private labels = AccountController.state.addressLabels
  
  private balance = 0
  
  private fetchingBalance = true
  
  private shouldShowIcon = false
  
  private caipNetwork = NetworkController.state.caipNetwork
```

File: packages/ui/src/composites/wui-list-account/index.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  @property({ type: Boolean }) public selected = false
+
+  @property({ type: Function }) public onSelect?: (
+    { address, type }: { address: string; type: string },
+    selected: boolean
+  ) => void
+
+  handleClick = (event: Event) => {
```
Comment (enesozturk) [inline packages/ui/src/composites/wui-list-account/index.ts:40]: Same for the anonymous functions and making it private

File: packages/ui/src/composites/wui-list-account/index.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    if (!label && this.connectedConnector === 'AUTH') {
+      label = `${this.accountType === 'eoa' ? 'Email' : 'Smart'} Account`
+    } else if (
+      (!label && this.connectedConnector === 'INJECTED') ||
+      this.connectedConnector === 'ANNOUNCED'
+    ) {
+      label = `Injected Account`
+    } else if (!label) {
+      label = 'EOA'
+    }
```
Comment (enesozturk) [inline packages/ui/src/composites/wui-list-account/index.ts:74]: Can we separate this logic into another private function like `getLabel()` and just call it in the render? Main render function should be as simple as possible to make the code more modular and maintainable.

File: packages/ui/src/composites/wui-list-account/styles.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import { css } from 'lit'
+
+export default css`
+  button {
+    padding: 6.5px var(--wui-spacing-l) 6.5px var(--wui-spacing-xs);
```
Comment (enesozturk) [inline packages/ui/src/composites/wui-list-account/styles.ts:5]: I know this is more of design question but just to double check if we are okay to use some value that is not in our spacing system. (I mean `6.5px`) We can skip this if it's a designers choice bc we generally want to stick to the design variables

File: packages/ui/src/composites/wui-profile-button-v2/index.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+              truncate: this.profileName ? 'end' : 'middle'
+            })}
+          </wui-text>
+          <wui-icon size="sm" color="fg-200" name="copy" id="copy-address"></wui-icon>
+        </wui-flex>
+      </wui-flex>
+    </button>`
+  }
+
+  private handleClick(event: Event) {
```
Comment (enesozturk) [inline packages/ui/src/composites/wui-profile-button-v2/index.ts:62]: Should be moved under the `// -- Private ---...` comment

**Review by enesozturk — CHANGES_REQUESTED** [review]

File: packages/core/src/controllers/OptionsController.ts [hunk 14]
```diff
@@ -21,6 +21,8 @@ export interface OptionsControllerState {
   enableAnalytics?: boolean
   metadata?: Metadata
   enableOnramp?: boolean
+  enableWalletFeatures?: boolean
```
[the remark this answers is not in the document]
Comment (ganchoradkov) [reply packages/core/src/controllers/OptionsController.ts:24]: fixed, seems to be result from the 10s of conflicts we had updating to latest

File: packages/ethers/src/client.ts [hunk 15]
```diff
@@ -772,14 +799,15 @@ export class Web3Modal extends Web3ModalScaffold {
     window?.localStorage.setItem(EthersConstantsUtil.WALLET_ID, name)
 
     if (provider) {
-      const { address, chainId } = await EthersHelpersUtil.getUserInfo(provider)
-      if (address && chainId) {
+      const { addresses, chainId } = await EthersHelpersUtil.getUserInfo(provider)
+      if (addresses?.[0] && chainId) {
```
[the remark this answers is not in the document]
Comment (ganchoradkov) [reply packages/ethers/src/client.ts:803]: updated, all check now for at least 1 account

File: packages/ethers/src/client.ts [hunk 16]
```diff
@@ -837,10 +867,10 @@ export class Web3Modal extends Web3ModalScaffold {
   private async watchWalletConnect() {
     const provider = await this.getWalletConnectProvider()
 
-    function disconnectHandler() {
+    const disconnectHandler = () => {
```
[the remark this answers is not in the document]
Comment (ganchoradkov) [reply packages/ethers/src/client.ts:870]: reverted 👍

File: packages/scaffold-ui/src/modal/w3m-router/index.ts (hunk 1 above)
Comment (ganchoradkov) [reply packages/scaffold-ui/src/modal/w3m-router/index.ts:93]: merge conflict, fixed

File: packages/scaffold-ui/src/partials/w3m-account-default-widget/index.ts (hunk 2 above)
Comment (ganchoradkov) [reply packages/scaffold-ui/src/partials/w3m-account-default-widget/index.ts:89]: I would leave this to be addressed at a later point such as tech debt week

File: packages/ui/src/composites/wui-dropdown-menu/styles.ts (hunk 8 above)
Comment (ganchoradkov) [reply packages/ui/src/composites/wui-dropdown-menu/styles.ts:13]: @tomiir will be best to answer

File: packages/ui/src/composites/wui-list-account/styles.ts (hunk 12 above)
Comment (ganchoradkov) [reply packages/ui/src/composites/wui-list-account/styles.ts:5]: copied it from similar components, this value (6.5px) is used on quite few places but can update ofc if required

File: packages/scaffold-ui/src/views/w3m-account-settings-view/styles.ts [hunk 17]
```diff
@@ -1,50 +1,3 @@
...
-  .account-links wui-flex wui-icon {
-    width: var(--S, 20px);
-    height: var(--S, 20px);
-  }
-
-  .account-links wui-flex wui-icon svg path {
-    stroke: #47a1ff;
-  }
-`
+export default css``
```
Comment (tomiir) [inline packages/scaffold-ui/src/views/w3m-account-settings-view/styles.ts:3]: If this file is no longer used I'd remove it directly. 
btw where did all of these styles go?

File: packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+  private selectedAccounts: AccountType[] = AccountController.state.allAccounts
+  private selectAll = true
+  private approved = false
+
+  @state() private isApproving = false
+  constructor() {
+    super()
+    AccountController.subscribeKey('allAccounts', allAccounts => {
+      this.allAccounts = allAccounts
+      this.requestUpdate()
```
Comment (tomiir) [inline packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts:30]: is this forced update required? have you tried making `allAccounts` a `@state` variable so that changes trigger a re-remder?

File: packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+  private onSelect = (account: AccountType, add: boolean) => {
+    if (add) {
+      this.selectedAccounts.push(account)
+    } else {
+      this.selectedAccounts = this.selectedAccounts.filter(a => a.address !== account.address)
+    }
+    if (this.selectedAccounts.length > 0) {
+      this.selectAll = this.selectedAccounts.length === this.allAccounts.length
+    }
+    this.requestUpdate()
```
Comment (tomiir) [inline packages/scaffold-ui/src/views/w3m-select-addresses-view/index.ts:108]: same as before, using a state variable would automatically trigger the render

File: packages/scaffold-ui/src/views/w3m-switch-address-view/index.ts [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+import { UiHelperUtil, customElement } from '@web3modal/ui'
+import { LitElement, html } from 'lit'
+import styles from './styles.js'
+
+@customElement('w3m-switch-address-view')
+export class W3mSwitchAddressView extends LitElement {
+  public static override styles = styles
+  // -- Members ------------------------------------------- //
+  private readonly metadata = OptionsController.state.metadata
+  public allAccounts: AccountType[] = AccountController.state.allAccounts || []
```
Comment (tomiir) [inline packages/scaffold-ui/src/views/w3m-switch-address-view/index.ts:19]: same as before I'd make this a state var

Comment (tomiir) [reply packages/scaffold-ui/src/views/w3m-switch-address-view/index.ts:19]: and probably a couple others to prevent manual re-renders

File: packages/ui/src/composites/wui-dropdown-menu/styles.ts (hunk 8 above)
Comment (tomiir) [reply packages/ui/src/composites/wui-dropdown-menu/styles.ts:13]: Doesn't seem to be a standardized shadow in the ThemeUtil. We use the same shadow on the snackbar so we could maybe abstract it but don't think it's a big deal. If we do I'd do it as part of design system variable review in general which we could do on tech debt week imo

File: packages/ui/src/composites/wui-list-account/index.ts [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+  public override connectedCallback() {
+    super.connectedCallback()
+    BlockchainApiController.getBalance(this.accountAddress, this.caipNetwork?.id).then(response => {
+      let total = this.balance
+      if (response.balances.length > 0) {
+        total = response.balances.reduce((acc, balance) => acc + (balance?.value || 0), 0)
+      }
+      this.balance = total
+      this.fetchingBalance = false
+      this.requestUpdate()
```
Comment (tomiir) [inline packages/ui/src/composites/wui-list-account/index.ts:55]: same as before I'd use state vars

**Review by tomiir — COMMENTED** [review]
Some comments, mostly interested in the state variables. Else looks ok!

File: packages/ui/src/composites/wui-list-account/styles.ts (hunk 12 above)
Comment (enesozturk) [reply packages/ui/src/composites/wui-list-account/styles.ts:5]: @ganchoradkov Then I guess we could keep it I'd like to not block your time for UI details. We should review those in general

File: packages/ui/src/composites/wui-list-account/index.ts (hunk 21 above)
Comment (ganchoradkov) [reply packages/ui/src/composites/wui-list-account/index.ts:55]: updated it but dangerjs complains https://github.com/WalletConnect/web3modal/pull/2286#issuecomment-2208322667

File: packages/wallet/src/W3mFrameSchema.ts [hunk 22]
```diff
@@ -84,6 +93,15 @@ export const FrameGetUserResponse = z.object({
...
   smartAccountDeployed: z.optional(z.boolean()),
+  accounts: z.array(
+    z.object({
+      address: z.string(),
+      type: z.enum([
+        W3mFrameRpcConstants.ACCOUNT_TYPES.EOA,
+        W3mFrameRpcConstants.ACCOUNT_TYPES.SMART_ACCOUNT
+      ])
+    })
+  ),
```
Comment (tomiir) [inline packages/wallet/src/W3mFrameSchema.ts:104]: doesn't this break secure site ?

**Review by tomiir — APPROVED** [review]

File: packages/wallet/src/W3mFrameSchema.ts (hunk 22 above)
Comment (ganchoradkov) [reply packages/wallet/src/W3mFrameSchema.ts]: updated with `optional()`
